### ISSUE TITLE
pounce.jax.solve backward treats inactive inequality constraints as equalities → incorrect dx*/dp

### ISSUE THREAD

**Issue by robertojjx** [jkitchin/pounce#73]
## Summary

`pounce.jax.solve`'s backward (the implicit-function-theorem `custom_vjp`) keeps **inactive inequality constraints** in the KKT system as equalities. As a result, `jax.grad` / `jax.jacobian` through the solve returns an **incorrect** `dx*/dp` whenever a general inequality constraint `cl <= g_i(x) <= cu` is *slack* at the optimum. The forward solve is correct; only the differentiation is affected. No error is raised — the wrong gradient is returned silently.

The active-set logic in the backward is applied to **variable bounds** (`mult_x_L` / `mult_x_U`) but **not** to **constraint** rows (`mult_g`).

## Where (POUNCE 0.2.0, `python/pounce/jax/_diff.py`)

In `bwd` ([permalink](https://github.com/jkitchin/pounce/blob/6be3643aa8d3c4e0446d56c5dba03802298c52a0/python/pounce/jax/_diff.py)):

- L123 — the working set is computed from **variable-bound** multipliers only:
  ```python
  active = (mult_xL > _ACTIVE_TOL) | (mult_xU > _ACTIVE_TOL)   # length n, over variables
  ```
- L139 — the **full** constraint Jacobian is built (all `m` rows):
  ```python
  J = jax.jacrev(g, argnums=0)(x_star, p)
  ```
- L157 — masking zeros active *variable columns*, never *constraint rows*:
  ```python
  J_eff = jnp.where(active[None, :], 0.0, J)
  ```
- L163–165 — every constraint row goes into the KKT block as an equality:
  ```python
  bot = jnp.concatenate([J_eff, jnp.zeros((m, m))], axis=1)
  ...
  rhs = jnp.concatenate([v_eff, jnp.zeros(m)])
  ```

`mult_g` is read into `lam` (L113) but used only to form the Lagrangian Hessian — never to select the constraint working set. So a slack inequality (with `mult_g_i ≈ 0`) is still linearized as `∇g_i · dx = 0`, over-constraining the sensitivity.

The implicit function theorem requires only the **active** set in the sensitivity system; including an inactive inequality pins the perturbed solution to a surface it isn't on.

## Documentation mismatch

- `docs/src/python.md`: "the gradient supplied by the implicit function theorem applied to the KKT conditions" — stated generally, no caveat.
- `python/notebooks/03_implicit_differentiation.ipynb` (Notes): "Active inequality bounds at the solution are handled in the backward..." — this describes **variable bounds**, but reads as if general **inequality constraints** are handled. They are not.

## Minimal reproduction

`min ||x - p||²` s.t. `x[0] <= 0.5` (a single one-sided inequality). Compare the analytic Jacobian (`jax.jacobian` through `pounce.jax.solve`) against a central finite difference.

```python
import numpy as np, jax, jax.numpy as jnp
jax.config.update("jax_enable_x64", True)
from pounce.jax import solve

N, B = 3, 0.5
def f(x, p): d = x - p; return jnp.dot(d, d)
def g(x, p): return jnp.stack([x[0]])            # x[0] <= B via cu

def forward(p):
    return solve(p, f=f, g=g, x0=jnp.zeros(N), n=N, m=1,
                 lb=jnp.full(N, -1e19), ub=jnp.full(N, 1e19),
                 cl=jnp.array([-1e19]), cu=jnp.array([B]),
                 options={"tol": 1e-10, "print_level": 0})

def fd(p, eps=1e-6):
    p = np.asarray(p, float); J = np.zeros((N, N))
    for j in range(N):
        e = np.zeros(N); e[j] = eps
        J[:, j] = (np.asarray(forward(jnp.asarray(p+e))) -
                   np.asarray(forward(jnp.asarray(p-e)))) / (2*eps)
    return J

# INACTIVE: p[0] = -1 < B  =>  x* = p  =>  true dx*/dp = I
p = jnp.array([-1.0, 2.0, -3.0])
print("analytic:\n", np.asarray(jax.jacobian(forward)(p)))
print("finite-diff (truth):\n", fd(p))
```

### Output (POUNCE 0.2.0, built `maturin develop --release`)

```
Case A: inequality INACTIVE (slack), p = [-1, 2, -3]   x* = p, truth dx*/dp = I
  analytic dx*/dp:        finite-diff (truth):
  [[-0. -0. -0.]           [[1. 0. 0.]
   [ 0.  1.  0.]            [0. 1. 0.]
   [ 0.  0.  1.]]           [0. 0. 1.]]
  max|analytic - FD| = 1.0     MATCHES FD: NO  <-- BUG

Case B: inequality ACTIVE (binding), p = [2, 2, -3]
  analytic dx*/dp matches FD to 1.4e-10              MATCHES FD: YES
```

The backward is correct when the inequality is active and wrong when it is slack — confirming the slack row is being kept as an equality.

## Suggested fix

Build a constraint working set from `mult_g`, exactly as the code already does for variable bounds: for any constraint row `i` with `|mult_g_i| <= _ACTIVE_TOL`, drop it from the KKT system (zero its row/column and identity-augment so the block stays invertible), mirroring the existing variable-bound handling.

## Context

Found while evaluating `pounce.jax` as the inner solver for a differentiable constrained-projection layer (project `y_hat` onto `c(y)=0, h(y)<=0` and backprop `dy*/dy_hat`). Equality- and bound-only problems differentiate correctly; the issue appears only with inequality constraints that are slack at the solution. Environment: macOS arm64, Python 3.10, jax 0.6.2, POUNCE 0.2.0 at `6be3643`.

**Comment by jkitchin** [comment]
Thanks for the precise repro and the source-line pointers — both made this a quick fix.

Fixed in b77af74 (on `main`).

### What changed

`bwd` in `python/pounce/jax/_diff.py` now builds a working set over **constraint rows** in addition to variable bounds:

```python
cl_arr = jnp.asarray(cl, dtype=H.dtype)
cu_arr = jnp.asarray(cu, dtype=H.dtype)
is_equality = cl_arr == cu_arr
cons_active = is_equality | (jnp.abs(lam) > _ACTIVE_TOL)
cons_inactive = ~cons_active
```

Equalities (`cl[i] == cu[i]`) are always active; inequalities are active iff their multiplier binds. Inactive rows are zeroed in `J_eff` and the bottom-right zero block of the KKT matrix is replaced by `diag(cons_inactive)` — same identity-augment trick already used for active variable bounds. Each slack row then reads `1 · u_lam[i] = 0` and decouples cleanly.

### Verifying with your repro

Running your script verbatim (extended to print both cases) against `main` now gives:

```
Case A: inequality INACTIVE (slack), p = [-1.0, 2.0, -3.0]
  analytic dx*/dp:
    [ 1. -0. -0.]
    [-0.  1. -0.]
    [-0. -0.  1.]
  finite-diff (truth):
    [1. 0. 0.]
    [0. 1. 0.]
    [0. 0. 1.]
  max|analytic - FD| = 1.40e-10    MATCHES FD: YES

Case B: inequality ACTIVE (binding), p = [2.0, 2.0, -3.0]
  analytic dx*/dp:
    [-0. -0. -0.]
    [ 0.  1.  0.]
    [ 0.  0.  1.]
  finite-diff (truth):
    [0. 0. 0.]
    [0. 1. 0.]
    [0. 0. 1.]
  max|analytic - FD| = 1.40e-10    MATCHES FD: YES
```

**Why this is right:**

- *Case A (slack):* p[0] = -1 < B = 0.5, so the inequality x[0] <= 0.5 is not binding at the optimum. The unconstrained minimizer of ||x - p||² is x* = p, so dx*/dp = I. The previous bug pinned dx*[0]/dp to zero by carrying the slack row as an equality; now that row is dropped from the KKT block and the identity is recovered to ~1.4e-10 (FD-vs-analytic at this eps).
- *Case B (binding):* p[0] = 2 > B, so x*[0] = 0.5 regardless of p[0], while x*[1:] = p[1:]. The top row of dx*/dp is zero; the rest is identity. This case was correct before and remains correct — confirming the fix doesn't regress the active path.

### Regression coverage

`python/tests/test_jax.py` now has `test_implicit_diff_inactive_inequality_pounce_73` and `test_implicit_diff_active_inequality_pounce_73` pinning both cases of your repro against finite-diff. I confirmed the INACTIVE test fails on 6be3643 (pre-fix) with the exact symptom you described and passes on b77af74.

### Where this lands in versions

`pounce-solver 0.3.0` (about to be tagged) will be the first PyPI release with this fix. If you're building from source on macOS arm64, `cd python && maturin develop --release` against `main` picks it up immediately.

Please give it a spin against your differentiable-constrained-projection layer (project y_hat onto c(y)=0, h(y)<=0) and reopen if anything still looks off — particularly if you exercise the two-sided inequality case (cl[i] = a, cu[i] = b with a < b), since that's lightly tested in the new fixtures.

**Comment by jkitchin** [comment]
Resolved by b77af74 (`fix(jax): respect constraint active set in pounce.jax.solve backward`).

The backward now builds a constraint working set from `mult_g` (`cons_active = is_equality | (|mult_g| > _ACTIVE_TOL)`), zeroing inactive constraint rows in `J_eff` and replacing the bottom-right zero block with `diag(cons_inactive)` so each slack row reads `1·u_lam[i] = 0`. Applied across all three backward paths in `python/pounce/jax/_diff.py`.

Regression tests `test_implicit_diff_{inactive,active}_inequality_pounce_73` pin the issue's repro against finite-diff and both pass. Docs and notebook 03 updated to distinguish variable-bound vs. constraint-row handling.
